### PR TITLE
Simplify decimal cast logic

### LINKED ISSUE #3265: Decimal128 to Decimal256 Overflows
**Describe the bug**

```
#[test]
    fn test_cast_decimal128_to_decimal256_negative() {
        let input_type = DataType::Decimal128(10, 3);
        let output_type = DataType::Decimal256(10, 5);
        assert!(can_cast_types(&input_type, &output_type));
        let array = vec![Some(i128::MAX), Some(i128::MIN)];
        let input_decimal_array = create_decimal_array(array, 10, 3).unwrap();
        let array = Arc::new(input_decimal_array) as ArrayRef;

        let hundred = i256::from_i128(100);
        generate_cast_test_case!(
            &array,
            Decimal256Array,
            &output_type,
            vec![
                Some(i256::from_i128(i128::MAX).mul_wrapping(hundred)),
                Some(i256::from_i128(i128::MIN).mul_wrapping(hundred))
            ]
        );
    }
```

Fails with

```
thread 'cast::tests::test_cast_decimal128_to_decimal256_negative' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `17014118346046923173168730371588410572700`', arrow-cast/src/cast.rs:7251:9
```

The issue is that it is not performing the scaling operation in the destination type

```
let iter = array.iter().map(|v| {
    v.and_then(|v| v.mul_checked(mul).ok().map(i256::from_i128))
});
let casted_array = unsafe {
    PrimitiveArray::<Decimal256Type>::from_trusted_len_iter(iter)
};
casted_array
    .with_precision_and_scale(*output_precision, *output_scale)
    .map(|a| Arc::new(a) as ArrayRef)
```

**To Reproduce**

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #3265

# Rationale for this change
 

The logic was extremely complicated and hard to follow, and suffered from a lot of duplication, leading to inconsistencies such as #3224 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
This does represent a performance regression for decimal256 arrays, as previously the logic was incorrectly using i128 arithmetic when it should have been using i256.

**Review by viirya — APPROVED** [review]
Nice!

**Comment by liukun4515** [comment]
> This does represent a performance regression for decimal256 arrays, as previously the logic was incorrectly using i128 arithmetic when it should have been using i256. Correctness is more important and we're talking microseconds, so I think this is fine

thanks @tustvold 
correctness is more important than performance.
